kernel/src/hil/screen.rs

```diff
@@ -186,18 +224,41 @@ pub trait Screen {
     /// Note that width and height may change due to rotation.
     fn get_resolution(&self) -> (usize, usize);
 
-    /// Returns the current pixel format
+    /// Returns the current pixel format.
+    ///
+    /// Some displays are limited to updating multiple pixels at a time.
+    /// Monochrome displays often accept packets of 8 bytes,
+    /// others need entire lines.
+    ///
+    /// This function returns the minimal shape that must be updated,
+    /// and its possible placement, as the `Grid` return value.
+    /// See `set_write_frame`.
+    ///
+    /// The driver MUST make sure that when the smallest square
+    /// described by the return value is filled with a pixel buffer,
+    /// then the buffer ends on a byte boundary.
+    /// For example, if the format is `Mono_1H8` (8 pixels per byte),
+    /// then `Grid::width` cannot be 12 (12 pixels = 12 bits = 1,5 bytes).
+    ///
     /// This function is synchronous as the driver should know this value without
-    /// requesting it from the screen.
-    fn get_pixel_format(&self) -> ScreenPixelFormat;
+    /// requesting it from the hardware.
+    fn get_pixel_format(&self) -> (PixelStreamFormat, Grid);
```

Comment:
I'm not exactly sure where the user should get that info. Grid might change with resolution, perhaps rotation, but returning a new Grid on every call is a bit silly.

Reply:
My goal here is to push the Grid information on the user, so that they don't have an opportunity not to notice it. After all, there's no way to operate the display without knowing all of format, resolution, and grid.

The idea to put it in get_pixel_format return value came from the fact that for a screen with only one supported format, get_pixel_format *must* be called before knowing how to submit data. (So does get_resolution, but I decided by a coin flip.)

I would like to preserve that property: shove the Grid in user's face. Does this make more sense now?

The situation gets more complicated in advanced displays where the user selects the pixel format. There, the user can change the pixel format directly, wihout a call to get_pixel_format. The path to push the Grid on to the user is broken. But I think that might have a solution:

```
fn set_pixel_format(Format) -> Result<Grid, Error>
```

(I think it's enough to put it in the return value rather than a callback, the driver will probably know how to calculate the new grid.) This way, the user will see the grid before submitting data.

Do this sound like a good idea?

Reply:
I think the `set_pixel_format` needs to be asynchronous, as many displays will have to ask this.

Reply:
Do you mean get_pixel_format? Are there any disadvantages of making the query at initialization?